File: 图论/最小生成树/py/kruskal.py
class UF:
    def __init__(self, n):
        self.n = n
        self.fa = [i for i in range(n)]
        self.sz = [1] * n
        self.comp_cnt = n

    def find(self, x):
        if x != self.fa[x]:
            self.fa[x] = self.find(self.fa[x])
        return self.fa[x]

    def unite(self, x, y):
        x, y = self.find(x), self.find(y)
        if x != y:
            if self.sz[x] < self.sz[y]:
                x, y = y, x
            self.fa[y] = x
            self.sz[x] += self.sz[y]
            self.comp_cnt -= 1

    def connected(self, x, y):
        return self.find(x) == self.find(y)

class Solution:
    def minimumCost(self, n: int, connections: List[List[int]]) -> int:
        edge = [(w, x-1, y-1) for x, y, w in connections]
        edge.sort()

        res = 0
        uf = UF(n)
        for w, x, y in edge:
            if uf.connected(x, y):
                continue
            uf.unite(x, y)
            res += w
            if uf.comp_cnt == 1:
                return res
        return -1

File: 图论/最小生成树/cpp/kruskal.cpp
// 稀疏图
const int INF = 0x3f3f3f3f, N = 1e5 + 10, M = 2e5 + 10;
int n, m;
int p[N];
struct Edge
{
	int a, b, w;

	bool operator<(const Edge& E) const
	{
		return w < E.w;
	}
} edges[M];
int find(int x)
{
	if (x != p[x]) p[x] = find(p[x]);
	return p[x];
}
void kruskal()
{
	int cnt = 0, res = 0;
	for (int i = 0; i < m; ++i)
	{
		int a = edges[i].a, b = edges[i].b, w = edges[i].w;
		int pa = find(a), pb = find(b);
		if (pa != pb)
		{
			p[pa] = pb;
			cnt++;
			res += w;
		}
	}
	if (cnt < n - 1) cout << "impossible" << endl;
	else cout << res << endl;
}
void solve()
{
	cin >> n >> m;
	for (int i = 1; i <= n; ++i) p[i] = i;
	for (int i = 0; i < m; ++i) cin >> edges[i].a >> edges[i].b >> edges[i].w;
	sort(edges, edges + m);
	kruskal();
}